core/base/inc/TObject.h

```diff
@@ -163,6 +163,16 @@ class TObject {
    virtual Int_t       Write(const char *name=0, Int_t option=0, Int_t bufsize=0);
    virtual Int_t       Write(const char *name=0, Int_t option=0, Int_t bufsize=0) const;
 
+   /// IsDestructed
+   ///
+   /// \note This function must be non-virtual as it can be used on destructed (but
+   /// not yet modified) memory.  This is used for example in TClonesArray to record
+   /// the element that have been destructed but not deleted and thus are ready for
+   /// re-use (by operator new with placement).
+   ///
+   /// \return true if this object's destructor has been run.
+   Bool_t IsDestructed() const { return ! TestBit(kNotDeleted); }
```

Comment:
I intentionally left it here as this function has no undefined behavior as it is intended to be use only on live memory (not deleted) to know whether the destructor was run on this memory or not.

Reply:
Technically true.  However since we are in control of the `TObject` code, we know how the memory is modified during the `TObject` `destruction` and any additional modification done implicit by compiler inserted code seems a tad unlikely (the memory is not released and so can not be reused yet) ; I suppose a tool like ASAN could instrument the destructors to taint the memory to try to warn about `use-after-destruction-but-before-deletion` but hopefully that would be optional, as without this feature (i.e. `IsDestructed` being predictable on un-released memory) `TClonesArray` is broken.